### PR TITLE
backends/ebpf: Change packet length calc to pass kernel verifier (#4158)

### PR DESCRIPTION
**Pull request by thomascalvert-xlnx**
This basically copies the strategy used in the PSA sub-backend.

Edit: as discovered in a comment below, this PR is only needed for passing the verifier in kernel versions prior to 5.9.

### CODE REVIEW

**Review by osinstom — APPROVED** [review]

**Comment by mihaibudiu** [comment]
Can you add a test as well? Thank you

**Comment by thomascalvert-xlnx** [comment]
@mihaibudiu could you please elaborate? As mentioned in the linked issue there is already a test, the failure of which prompted this fix. Are you looking for something else?
```
sudo ctest --output-on-failure -R 'ebpf-kernel/testdata/p4_16_samples/length_ebpf.p4'
```

**Comment by mihaibudiu** [comment]
We usually want a test that would fail without this patch and will pass with the patch.
It's just regression testing, to make sure this bug is not reintroduced in later commits.
The fact that nothing has changed under testdata (or no xfail was removed) indicates that there is no such test part of the PR.

**Comment by jafingerhut** [comment]
Perhaps the failing test that Thomas is pointing out that he fixed is not being run in any CI test currently?

**Comment by mihaibudiu** [comment]
Making it be part of the CI would be the right approach

**Comment by mihaibudiu** [comment]
I also think you need to run clang-format on your code

**Comment by thomascalvert-xlnx** [comment]
> Making it be part of the CI would be the right approach

@mihaibudiu @jafingerhut I think the reason this test (and others) are not run as part of CI is because they require `root` access in order to configure the network namespaces and load the eBPF program into the kernel. https://github.com/p4lang/p4c/blob/6e7cc248043469740de79d78a788424303879ed4/backends/ebpf/targets/kernel_target.py#L182-L184  From [this stackoverflow](https://stackoverflow.com/a/65956131) answer it appears that it may be possible to load eBPF programs as a user but I think this would require significant rework of the test harness (and doesn't handle the namespaces part).

> I also think you need to run clang-format on your code

Yes apologies I'm still ramping up with this part of the process. Will fix the formatting issues here & in the other PR.

**Comment by fruffy** [comment]
> > Making it be part of the CI would be the right approach
> 
> @mihaibudiu @jafingerhut I think the reason this test (and others) are not run as part of CI is because they require `root` access in order to configure the network namespaces and load the eBPF program into the kernel.
> 
> https://github.com/p4lang/p4c/blob/6e7cc248043469740de79d78a788424303879ed4/backends/ebpf/targets/kernel_target.py#L182-L184
> From [this stackoverflow](https://stackoverflow.com/a/65956131) answer it appears that it may be possible to load eBPF programs as a user but I think this would require significant rework of the test harness (and doesn't handle the namespaces part).
> 
> > I also think you need to run clang-format on your code
> 
> Yes apologies I'm still ramping up with this part of the process. Will fix the formatting issues here & in the other PR.

Just looked at this. Actually, the reason is that the libbpf dependency is built after we check for its existence... Let me fix this.

**Comment by thomascalvert-xlnx** [comment]
Fixed clang-format issues and rebased onto top of main.

**Comment by thomascalvert-xlnx** [comment]
Thanks @fruffy for providing a fix. Indeed I was wrong and CI tests are already run as root.

@jafingerhut the failures in #4166 seem to be different to this one - the errors are about a relocation without PIE. Surprisingly, the `length_ebpf` test seems to have passed in that regression run. My only guess at this time is that it could be due to different kernel versions, `5.15.0-1046-azure` vs `5.4.0-144-generic`.

After some more digging I found [this patch](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/kernel/bpf/verifier.c?id=6c6935419e2fd8ef1fcde71c4e50b9095e520900), which seems to suggest that the kernel verifier accepts this construct as of kernel 5.9. This is consistent with the observed behaviour.

In summary: this patch only applies to kernels versions prior to 5.9. At a glance, the CI runs all seem to be running more recent kernels (Ubuntu20: `5.15.0-1046-azure`; all others: `6.2.0-1011-azure`)

**Comment by thomascalvert-xlnx** [comment]
This got included in https://github.com/p4lang/p4c/commit/aefe41fae5c381644262ba318df03124615c04be - closing.
